server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java

```diff
@@ -44,28 +47,37 @@
 
     /** default ctor primarily used for tests without plugins */
     public EngineConfigFactory(IndexSettings idxSettings) {
-        this(Collections.emptyList(), idxSettings);
+        this(Collections.emptyList(), idxSettings, null);
     }
 
     /**
      * Construct a factory using the plugin service and provided index settings
      */
     public EngineConfigFactory(PluginsService pluginsService, IndexSettings idxSettings) {
-        this(pluginsService.filterPlugins(EnginePlugin.class), idxSettings);
+        this(pluginsService.filterPlugins(EnginePlugin.class), idxSettings, null);
+    }
+
+    /**
+     * Construct a factory using the plugin service, provided index settings and mapper service
+     */
+    public EngineConfigFactory(PluginsService pluginsService, IndexSettings idxSettings, MapperService mapperService) {
+        this(pluginsService.filterPlugins(EnginePlugin.class), idxSettings, mapperService);
     }
 
     /* private constructor to construct the factory from specific EnginePlugins and IndexSettings */
-    EngineConfigFactory(Collection<EnginePlugin> enginePlugins, IndexSettings idxSettings) {
+    EngineConfigFactory(Collection<EnginePlugin> enginePlugins, IndexSettings idxSettings, MapperService mapperService) {
         Optional<CodecService> codecService = Optional.empty();
         String codecServiceOverridingPlugin = null;
+        Optional<CodecServiceFactory> codecServiceFactory = Optional.empty();
+        String codecServiceFactoryOverridingPlugin = null;
         Optional<TranslogDeletionPolicyFactory> translogDeletionPolicyFactory = Optional.empty();
         String translogDeletionPolicyOverridingPlugin = null;
         for (EnginePlugin enginePlugin : enginePlugins) {
             // get overriding codec service from EnginePlugin
             if (codecService.isPresent() == false) {
                 codecService = enginePlugin.getCustomCodecService(idxSettings);
                 codecServiceOverridingPlugin = enginePlugin.getClass().getName();
-            } else {
+            } else if (enginePlugin.getCustomCodecService(idxSettings).isPresent()) {
```

Comment:
This seems to be a bug: the `IllegalStateException` is going to be throw even if `enginePlugin` does not provide any codec service (returns `Optional.empty`).

Reply:
Seems bad. Has this case just never been hit because in practice there is only ever 1 engine plugin or because the last plugin in the list is the one to override the codec service?

Reply:
Hard to say, but I would guess that only handful of engine plugins do provided codec service (and probably there is usually only 1 as you mentioned).

Reply:
Right, and I believe this change was introduced in 1.2 as well